### PR TITLE
Delete Author

### PR DESCRIPTION
**Pull request by jmonteiro**
PT: https://www.pivotaltracker.com/story/show/126530671

Added ability to re-assign articles to new author & delete author.

# QA instructions

 1. Go to our staging environment.
 1. Click on Authors.
 1. Select an author with at least 1 contribution.
 1. You should see "Danger Zone" in the bottom of the page.
 1. Choose another user and delete the current author.
 1. New author should get all contributions (articles created, edited and marked as rot) from deleted author.

# Technical details

It's in the bottom of the Author's page, called "Danger Zone" (GitHub inspired). This section is only visible to administrators, which are specified in the .env file (didn't want to add a whole new permission system to Orientation just for that, but also didn't want to allow anyone to delete authors).

orientation

Confirmation prompt after clicking the DELETE AUTHOR button:

orientation

P.S.: due to pending PR #37, I am comparing this with `official_sync` branch for now, will change that to `pseudo-master` before merging.

### CODE REVIEW

**Comment by jmonteiro** [comment]
@brupm, @benmanns, @AhmedBelal, @DanielNill and @maletor: please review.

File: app/models/user.rb [hunk 1]
```diff
@@ -56,6 +57,10 @@ def self.text_search(query)
     end
   end
 
+  def administrator?
+    ENV['ADMINISTRATORS'].split(':').include? id.to_s
```
Comment (brupm) [inline app/models/user.rb]: why not use an array?

File: app/controllers/authors_controller.rb [hunk 2]
```diff
@@ -57,6 +57,23 @@ def toggle_email_privacy
...
+    @author = User.find(params[:id])
+
+    if !current_user.administrator?
+      flash[:notice] = "You are not an administrator."
+      redirect_to author_path(@author)
+    elsif current_user == @author
+      flash[:notice] = "You cannot delete yourself."
+      redirect_to author_path(@author)
+    else
+      replacement_author = User.find params[:replacement_author_id]
```
Comment (brupm) [inline app/controllers/authors_controller.rb]: `find()` keep params usage consistent

File: app/views/authors/show.html.haml [hunk 3]
```diff
@@ -67,9 +67,24 @@
...
+            = form_tag author_path( @author ), method: :delete, class: 'form' do
+              .form-field
+                = label_tag :replacement_author_id do
+                  %span.form-label.label
+                    Re-assign articles to
+                    = select_tag 'replacement_author_id', options_from_collection_for_select( User.active.order(:name), 'id', 'name', current_user.id )
+                    and
+                    = submit_tag "Delete Author",
+                      class: 'btn btn--a btn--1',
+                      data: { disable_with: 'Deleting...', confirm: "Articles created, edited or marked as rot by #{@author.name} will be re-assigned to the selected replacement author. Continue?" }
```
Comment (brupm) [inline app/views/authors/show.html.haml]: Message is a bit confusing, I would suggest: 

"Articles owned by #{author} will be re-assigned to selected user."

File: app/models/user.rb (hunk 1 above)
Comment (jmonteiro) [inline app/models/user.rb]: It's an ENV var (in our specific case, set through `heroku config:add`), and as such we should always threat is as a string. Although, for instance, BASH supports "array" variables, it's considered "buggy" ([source](http://stackoverflow.com/a/14100549))

File: app/views/authors/show.html.haml (hunk 3 above)
Comment (jmonteiro) [inline app/views/authors/show.html.haml]: Done.

File: app/controllers/authors_controller.rb (hunk 2 above)
Comment (jmonteiro) [inline app/controllers/authors_controller.rb]: Done.

File: app/models/user.rb (hunk 1 above)
Comment (benmanns) [inline app/models/user.rb]: :+1: but I might use commas for this.

File: app/models/user.rb [hunk 4]
```diff
@@ -88,6 +93,10 @@ def to_s
     self.name || self.email
   end
 
+  def replace_and_destroy!(other_user)
+    replace_with_user!(other_user) && destroy!
```
Comment (benmanns) [inline app/models/user.rb]: I would wrap this in a transaction.

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -88,4 +88,72 @@
...
+    let(:replacement) { create(:user) }
+    let(:article) { create(:article, author: user) }
+    let(:edit) { create(:article, editor: user) }
+    let(:rot_report) { create(:article, rot_reporter: user) }
+
+    it 'replaces with another user and destroys' do
+      expect do
+        expect do
+          expect do
+            expect(user.replace_and_destroy!(replacement)).to eq user
```
Comment (benmanns) [inline spec/models/user_spec.rb]: I wouldn't nest these this way. You could break into separate `it` blocks.

Comment (jmonteiro) [inline spec/models/user_spec.rb]: Done. 👍

File: app/models/user.rb (hunk 4 above)
Comment (jmonteiro) [inline app/models/user.rb]: Done. 🎉

**Comment by brupm** [comment]
👍

**Comment by SowjanyaEga** [comment]
Ready for deployment.

**Pull request merged**